service/migration_manager.cc

```diff
@@ -1159,29 +1159,43 @@ future<schema_ptr> migration_manager::get_schema_for_write(table_schema_version
     }
 
     auto s = local_schema_registry().get_or_null(v);
-
-    if (s && s->is_synced()) {
-        co_return s;
-    }
-
     // `_enable_schema_pulls` may change concurrently with this function (but only from `true` to `false`).
     bool use_raft = !_enable_schema_pulls;
-
-    if (use_raft) {
+    if ((!s || !s->is_synced()) && use_raft) {
         // Schema is synchronized through Raft, so perform a group 0 read barrier.
         // Batch the barriers so we don't invoke them redundantly.
         mlogger.trace("Performing raft read barrier because schema is not synced, version: {}", v);
         co_await (as ? _group0_barrier.trigger(*as) : _group0_barrier.trigger());
     }
 
-    s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
+    if (!s) {
+        s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
+    }
 
-    if (use_raft) {
-        // If Raft is used the schema is synced already (through barrier above), mark it as such.
-        mlogger.trace("Mark schema {} as synced", v);
-        co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
-    } else {
-        co_await maybe_sync(s, dst);
+    if (!s->is_synced()) {
+        if (use_raft) {
+            // If Raft is used the schema is synced already (through barrier above), mark it as such.
+            mlogger.trace("Mark schema {} as synced", v);
+            co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
+        } else {
+            co_await maybe_sync(s, dst);
+        }
+    }
```

Comment:
The refactors performed above should be in a separate commit, they are not really part of the fix, it would make the reviewer's life much easier.